utils/recipients.py

```diff
@@ -0,0 +1,218 @@
+    def rows_with_missing_data(self):
+        return set(
+            row['index'] for row in self.rows_annotated if any(
+                key not in [self.recipient_column_header, 'index'] and value.get('error')
+                for key, value in row.items()
+            )
+        )
+
+    @property
+    def rows_with_bad_recipients(self):
+        return set(
+            row['index'] for row in self.rows_annotated if row.get(self.recipient_column_header, {}).get('error')
+        )
+
+    @property
+    def rows_annotated(self):
+        for row_index, row in enumerate(self.rows):
+            yield dict(
+                {key: {
+                    'data': value,
+                    'error': self._get_error_for_field(key, value),
+                    'ignore': (key != self.recipient_column_header and key not in self.placeholders)
+                } for key, value in row.items()},
+                index=row_index
+            )
+
+    @property
+    def rows_annotated_and_truncated(self):
+        rows_with_errors = set()
+        for row in self.rows_annotated:
+            if RecipientCSV.row_has_error(row):
+                rows_with_errors.add(row['index'])
+            if len(rows_with_errors) > self.max_errors_shown:
+                return
+            if row['index'] >= self.max_initial_rows_shown and row['index'] not in rows_with_errors:
+                continue
+            else:
+                yield row
+
+    @property
+    def recipients(self):
+        for row in self.rows:
+            yield self._get_recipient_from_row(row)
+
+    @property
+    def personalisation(self):
+        for row in self.rows:
+            yield self._get_personalisation_from_row(row)
+
+    @property
+    def recipients_and_personalisation(self):
+        for row in self.rows:
+            yield (
+                self._get_recipient_from_row(row),
+                self._get_personalisation_from_row(row)
+            )
+
+    @property
+    def column_headers(self):
+        for row in csv.reader(
+            self.file_data.strip().splitlines(),
+            **RecipientCSV.reader_options
+        ):
+            return row
+        return []
+
+    @property
+    def missing_column_headers(self):
+        return set([self.recipient_column_header] + list(self.placeholders)) - set(self.column_headers)
+
+    @property
+    def column_headers_with_placeholders_highlighted(self):
+        return [
+            Markup(Template.placeholder_tag.format(header)) if header in self.placeholders else header
+            for header in self.column_headers
+        ]
+
+    def _get_error_for_field(self, key, value):
+        if key == self.recipient_column_header:
+            try:
+                validate_recipient(value, self.template_type)
+            except (InvalidEmailError, InvalidPhoneError) as error:
+                return str(error)
+        if key not in self.placeholders:
+            return None
+        if value in [None, '']:
```

Comment:
Good point. Looks like `csv` handles it, and I’ve [added a test](https://github.com/alphagov/notifications-utils/pull/9/files#diff-4e832f9d1c0b620621bf4a2ad14c83b6R185) just to make sure.